akka-http-core/src/main/scala/akka/http/scaladsl/ConnectionContext.scala

```diff
@@ -4,55 +4,158 @@
-import java.util.{ Optional, Collection => JCollection }
+import java.util.{ Collections, Optional, Collection => JCollection }
 
 import javax.net.ssl._
 
+import scala.collection.JavaConverters._
 import scala.collection.immutable
 import scala.compat.java8.OptionConverters._
+import scala.util.{ Failure, Success, Try }
+import com.typesafe.sslconfig.akka.AkkaSSLConfig
+import com.typesafe.sslconfig.akka.util.AkkaLoggerFactory
+import com.typesafe.sslconfig.ssl.DefaultHostnameVerifier
+import akka.actor.ClassicActorSystemProvider
+import akka.annotation.{ ApiMayChange, InternalApi }
+import akka.stream.{ ConnectionException, TLSClientAuth }
+import akka.stream.TLSProtocol._
 
 trait ConnectionContext extends akka.http.javadsl.ConnectionContext {
   @deprecated("Internal method, left for binary compatibility", since = "10.2.0")
   protected[http] def defaultPort: Int
 }
 
 object ConnectionContext {
-  // ConnectionContext
+  /**
+   *  If you want complete control over how to create the SSLEngine you can use this method:
+   *
+   *  Note that this means it is up to you to make sure features like SNI and Hostname Verification
+   *  are enabled as needed.
+   */
+  @ApiMayChange
+  def https(createSSLEngine: Option[(String, Int)] => () => SSLEngine): HttpsConnectionContext = // ...
+    new HttpsConnectionContext(Right(o => Engine(createSSLEngine(o))))
+
+  //#https-context-creation
+  /**
+   *  Use this sslContext to create a HttpsConnectionContext for server-side use.
+   */
+  def httpsServer(sslContext: SSLContext): HttpsConnectionContext = // ...
+    //#https-context-creation
+    new HttpsConnectionContext(Right({
+      case None => Engine(() => {
+        val engine = sslContext.createSSLEngine()
+        engine.setUseClientMode(false)
+        engine
+      })
+      case Some((host, port)) => Engine(() => {
+        val engine = sslContext.createSSLEngine(host, port)
+        engine.setUseClientMode(false)
+        engine
+      })
+    }))
+
   //#https-context-creation
+  /**
+   *  Use this sslContext to create a HttpsConnectionContext for client-side use.
+   */
+  def httpsClient(context: SSLContext)(implicit system: ClassicActorSystemProvider): HttpsConnectionContext = // ...
+    //#https-context-creation
+    {
+      val verifier = new DefaultHostnameVerifier(new AkkaLoggerFactory(system.classicSystem))
+
+      new HttpsConnectionContext(Right {
+        case None =>
+          Engine(() => {
+            val engine = context.createSSLEngine()
+            engine.getSSLParameters.setEndpointIdentificationAlgorithm("https")
+            engine.setUseClientMode(true)
+            engine
+          })
+        case Some((host, port)) =>
+          Engine(
+            () => {
+              val engine = context.createSSLEngine(host, port)
+              engine.getSSLParameters.setEndpointIdentificationAlgorithm("https")
+              engine.getSSLParameters.setServerNames(Collections.singletonList(new SNIHostName(host)))
```

Comment:
This is not needed. The SSLEngine will always add the hostname as given to `createSSLEngine(host, ...)` to the list of server names. Is it documented? You wish...